### DISCUSSION TITLE
Dynamically Created CRDs

### DISCUSSION THREAD

**Discussion by texasbobs** [fluxcd/flux2#2647] [General] [answered]
We are trying to use an app (OPA  Policy Gatekeeper).  Because we have 2200 clusters, we use a deployment tool to write to all 2200 of our repos that Flux consumes.  When we deploy an app to a cluster, the app in its entirety gets written to the target repo.

The problem is that some of the CRDs do not exist at deployment time.  In Flux, it would get to eventual consistency (CRs would get created after the CRD existed.)  Is there any way to do this with Flux2? 

In our case, we have a CR called constraintTemplate that dynamically creates a CRD called K8sPSPPrivilegedContainer. (Others would be created, too, as more constrainted get implemented.)

I don't think the `DependsOn` works for us because that means that the constraintTemplate and `K8sPSPPrivilegedContainer` CR have to be in two different directories and two different `kustomization`. 

```
{"level":"error","ts":"2022-04-18T17:51:50.480Z","logger":"controller.kustomization","msg":"Reconciliation failed after 5.712266175s, next try in 1h0m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"opa-gatekeeper","namespace":"flux-system","revision":"master/630416ee4058d2538d38dcc9e0e99b42550d446c","error":"K8sPSPPrivilegedContainer/psp-privileged-container dry-run failed, error: no matches for kind \"K8sPSPPrivilegedContainer\" in version \"constraints.gatekeeper.sh/v1beta1\"\n"
```

The two objects in question:

```
apiVersion: templates.gatekeeper.sh/v1beta1
kind: ConstraintTemplate
metadata:
  name: k8spspprivilegedcontainer
  annotations:
    description: >-
      Controls the ability of any container to enable privileged mode.
      Corresponds to the `privileged` field in a PodSecurityPolicy. For more
      information, see
      https://kubernetes.io/docs/concepts/policy/pod-security-policy/#privileged
spec:
  crd:
    spec:
      names:
        kind: K8sPSPPrivilegedContainer
      validation:
        openAPIV3Schema:
          description: >-
            Controls the ability of any container to enable privileged mode.
            Corresponds to the `privileged` field in a PodSecurityPolicy. For more
            information, see
            https://kubernetes.io/docs/concepts/policy/pod-security-policy/#privileged
  targets:
    - target: admission.k8s.gatekeeper.sh
      rego: |
        package k8spspprivileged
        violation[{"msg": msg, "details": {}}] {
            c := input_containers[_]
            c.securityContext.privileged
            msg := sprintf("Privileged container is not allowed: %v, securityContext: %v", [c.name, c.securityContext])
        }
        input_containers[c] {
            c := input.review.object.spec.containers[_]
        }
        input_containers[c] {
            c := input.review.object.spec.initContainers[_]
        }

apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sPSPPrivilegedContainer
metadata:
  name: psp-privileged-container
spec:
  match:
    kinds:
      - apiGroups: [""]
        kinds: ["Pod"]
    excludedNamespaces: ["kube-system"]
```

**Comment by rodrigoscferraz** [comment, marked as answer]
I guess this is what you are looking for
https://fluxcd.io/docs/components/helm/helmreleases/#crds

**Reply by texasbobs** [reply]
Doesn't look like a fit.  We're talking about CRDs that do not exist in a git repo.  So as the kustomization controller does a dry run, it fails because the CRD does not exist in the directory.  (This is because it is dynamically created.)

**Reply by tnowak-plex** [reply]
@texasbobs did you ever find a workaround for this?

We currently use gatekeeper and flux 0.17.2, which did not enforce dry-run.  Looking to upgrade to the latest version of flux, and this is a roadblock.

**Reply by texasbobs** [reply]
> @texasbobs did you ever find a workaround for this?
> 
> We currently use gatekeeper and flux 0.17.2, which did not enforce dry-run. Looking to upgrade to the latest version of flux, and this is a roadblock.

We also experienced this with the way Rancher deployed the System Upgrade Controller (dynamically created CRDs).  In the end, we split the CRs and the CRDs they need into different paths (and this way they are sync'ed by different flux `kustomization`s).  This means that the CRs will fail to be applied until the CRD(s) get created.  With this approach, we get eventual consistency; just not on the first pass.

**Reply by poblin-orange** [reply]
Yes, splitting CR definition and CR usage in 2 distinct fluxcd Kustomization is the way to go (quite typical, many operators can be installed with helm charts, further configuration must be in a distinct Kustomization to accomodate flux dry-run checks)
